src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,265 @@ title: Build System
+
+[Docker](https://www.docker.com/) is used to containerise our build processes so that we may mimic the build environment of the robot and impose version control over our required 3rd party dependencies while also managing operating system level configuration to ensure that all users of our codebase can achieve repeatable results.
+
+### NUClear Roles
+
+Full documentation for NUClear Roles can be found in its github [README](https://github.com/Fastcode/NUClearRoles/blob/master/README.md).
+
+Six main directories are used in a NUClear Roles system:
+
+| Directory        | Usage                  |
+| ---------------- | ---------------------- |
+| module           | NUClear reactors       |
+| roles            | Declaration of roles   |
+| shared/extension | NUClear DSL extensions |
+| shared/message   | Protobuf messages      |
+| shared/utility   | Utility code           |
+| tools            | Extensions for `b`     |
+
+#### Modules
+
+The modules folder contains [NUClear reactors](https://nuclear.readthedocs.io/en/latest/components.html#reactors). Reactors are grouped into meta-modules. For instance, the `vision` meta-module groups together reactors that perform vision processing tasks (e.g. ball and goal detection), while the `input` meta-module groups togehter reactors that gather system inputs (e.g. camera and network inputs).
+
+#### Roles
+
+Role files define which modules should be grouped together in order to form a single role (a binary that performs a specific role).
+
+Role files live in the `roles` folder and have a `.role` extension.
+
+Every role file must call the `nuclear_role()` function. Each argument to the `nuclear_role()` function is a C++ namespaced path to a module that should be included in the role. For example, if the `KinematicsConfiguration` module in `module/motion/KinematicsConfiguration` should be included in the role then `motion::KinematicsConfiguration` should appear as an argument to `nuclear_role()`.
+
+---
+
+**NOTE**
+`module` is removed from the start of all module arguments as all modules must reside in the `modules` folder.
+
+---
+
+Ideally, all modules should be independant of each other and, as such, it should not matter which order modules are specified to `nuclear_role()`. In reality, there are a couple of modules which, if they are included in a role, should be passed to `nuclear_role()` first. These modules are:
+
+- `extension::FileWatcher`
+- `support::SignalCatcher`
+- `support::logging::ConsoleLogHandler` or `support::logging::FileLogHandler`
+- `motion::KinematicsConfiguration`
+
+Below is an example of a role file. Any line preceeded by a `#` is a comment and will be ignored.
+
+```cmake
+nuclear_role(
+  # FileWatcher, ConsoleLogHandler and Signal Catcher Must Go First. KinematicsConfiguration usually goes after these
+  # and without it many roles do not run
+  extension::FileWatcher
+  # support::SignalCatcher
```

Comment:
SignalCatcher is commented out here, this might be confusing in the documentation